File: solutions/java/dnd-character/2/src/main/java/DnDCharacter.java
import java.math.BigDecimal;
import java.math.RoundingMode;
import java.util.stream.IntStream;

class DnDCharacter {
    static private int rollD6() {
        return Double.valueOf(Math.floor(Math.random() * 6) + 1).intValue();
    }

    static int ability() {
        return IntStream.of(DnDCharacter.rollD6(), DnDCharacter.rollD6(), DnDCharacter.rollD6(), DnDCharacter.rollD6())
            .sorted()
            .skip(1)
            .sum();
    }

    static int modifier(int input) {
        return BigDecimal.valueOf(Math.floor((input - 10) / 2f)).setScale(0, RoundingMode.FLOOR).intValue();
    }

    private int strength = DnDCharacter.ability();
    private int dexterity = DnDCharacter.ability();
    private int constitution = DnDCharacter.ability();
    private int intelligence = DnDCharacter.ability();
    private int wisdom = DnDCharacter.ability();
    private int charisma = DnDCharacter.ability();
    private int hitPoints = 10 + DnDCharacter.modifier(this.constitution);

    int getStrength() {
        return this.strength;
    }

    int getDexterity() {
        return this.dexterity;
    }

    int getConstitution() {
        return this.constitution;
    }

    int getIntelligence() {
        return this.intelligence;
    }

    int getWisdom() {
        return this.wisdom;
    }

    int getCharisma() {
        return this.charisma;
    }

    int getHitpoints() {
        return this.hitPoints;
    }

}

File: solutions/c/dnd-character/2/dnd_character.c
#include "dnd_character.h"

#include <stdio.h>
#include <stdbool.h>
#include <stdlib.h>
#include <time.h>

int ability(void){
    static bool didSeed = false;
    if (!didSeed){
        srand(time(NULL));
        didSeed = true;
    }
    
    char rolls[4] = {
        (rand() % 6) + 1,
        (rand() % 6) + 1,
        (rand() % 6) + 1,
        (rand() % 6) + 1,
    };

    int sum = 0, min_value = 7;
    for (int i = 0; i < 4; ++i){
        sum += rolls[i];
        if (rolls[i] < min_value){
            min_value = rolls[i];
        }
    }

    return sum - min_value;
}

int modifier(int score){
    return (score / 2) - 5;
}

dnd_character_t make_dnd_character(void){
    dnd_character_t result = (dnd_character_t){
        .strength = ability(),
        .dexterity = ability(),
        .constitution = ability(),
        .intelligence = ability(),
        .wisdom = ability(),
        .charisma = ability(),
        .hitpoints = 0
    };
    result.hitpoints = 10 + modifier(result.constitution);
    return result;
}